src/main/java/org/hydev/mcpm/server/crawlers/spiget/CreateDatabase.java

```diff
@@ -0,0 +1,186 @@
+package org.hydev.mcpm.server.crawlers.spiget;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.apache.hc.client5.http.utils.Hex;
+import org.hydev.mcpm.client.models.Database;
+import org.hydev.mcpm.client.models.PluginModel;
+import org.hydev.mcpm.client.models.PluginVersion;
+import org.hydev.mcpm.client.models.PluginYml;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.yaml.snakeyaml.error.MarkedYAMLException;
+
+import java.io.BufferedInputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * Responsible for statically generating database "index" files for mcpm.
+ * Used for searches, etc. To create a database use the createDatabase method
+ * with a path to the .mcpm/pkgs/spiget directory generated previously by SpigetCrawler.
+ */
+public class CreateDatabase {
+    public static final String packageStore = ".mcpm";
+
+    /**
+     * Main method generates a database in the mcpm directory.
+     *
+     * @param args Arguments are ignored.
+     */
+    public static void main(String[] args) {
+        writeDatabase(
+            new File(packageStore, "pkgs/spiget"),
+            new File(packageStore, "db"),
+            new File(packageStore, "db.hash")
+        );
+    }
+
+    /**
+     * Iterates through the crawlerDirectory and assembles and writes a database object to databaseFile.
+     *
+     * @param crawlerDirectory Path to the .mcpm/pkgs/spiget directory generated by SpigetCrawler.
+     * @param databaseFile Path to the db.json file that will be written to.
+     */
+    public static void writeDatabase(File crawlerDirectory, File databaseFile, @Nullable File versionFile) {
+        Database database = createDatabase(crawlerDirectory);
+
+        if (database == null)
+            return;
+
+        try {
+            new ObjectMapper().writeValue(databaseFile, database);
+
+            if (versionFile != null) {
+                try {
+                    var hash = generateHash(databaseFile);
+
+                    Files.writeString(versionFile.toPath(), hash);
+                } catch (NoSuchAlgorithmException | IOException e) {
+                    System.out.println("Failed to write database hash file.");
+                }
+            }
+        } catch (IOException e) {
+            e.printStackTrace();
+
+            System.out.println("Failed to write database file.");
+        } catch (Exception e) {
+            e.printStackTrace();
+
+            System.out.println("Failure in DB creation.");
+        }
```

Comment:
Hmm, ArchLinux actually has "lastupdate" and "lastsync" endpoints which give unix timestamps. I think it's because they have a huge mirror list and they need an automatic system to check which ones are maintained and which ones are outdated (since just having the hash doesn't tell you about update time, just tells you whether it is outdated).

So if you want, we can write a timestamp file along with the hash file. Should be an easy change.

<img width="333" alt="image" src="https://user-images.githubusercontent.com/22280294/199579100-df3a32e8-a140-4fc4-beeb-ae658d0ea928.png">

<img width="779" alt="image" src="https://user-images.githubusercontent.com/22280294/199578829-9d0f8264-95fe-4e01-80d8-9ecf79b543d6.png">

Reply:
Yeah I can do that :D 

What are we doing with the timestamp though? Does it matter if we have an outdated by 2 months file or an outdated by 1 month file?

Reply:
Maybe (if we have demand in the future) we can make an automatic mirror tracker that tracks mirrors' availability and update frequency, and calculates score for each mirror like https://archlinux.org/mirrors/status/.